File: C++/examples/cache_utils_example.cpp
/**
 * @file cache_utils_example.cpp
 * @brief Example usage of Cache Utilities
 * @version 1.0.0
 *
 * This file demonstrates various features of the Cache Utilities library:
 * - Basic cache operations
 * - TTL (Time To Live) expiration
 * - LRU eviction policy
 * - Statistics tracking
 * - Get or compute pattern
 * - Thread-safe operations
 */

#include <iostream>
#include <string>
#include <chrono>
#include <thread>
#include <vector>
#include "../cache_utils/mod.hpp"

using namespace alltoolkit;

// Example 1: Basic cache usage
void example_basic_usage() {
    std::cout << "\n=== Example 1: Basic Cache Usage ===" << std::endl;

    Cache<std::string, std::string> cache(100);

    // Store values
    cache.set("name", "Alice");
    cache.set("city", "New York");
    cache.set("country", "USA");

    // Retrieve values
    std::string value;
    if (cache.get("name", value)) {
        std::cout << "Name: " << value << std::endl;
    }

    if (cache.get("city", value)) {
        std::cout << "City: " << value << std::endl;
    }

    // Check if key exists
    std::cout << "Has 'name': " << (cache.has("name") ? "yes" : "no") << std::endl;
    std::cout << "Has 'age': " << (cache.has("age") ? "yes" : "no") << std::endl;

    // Remove a key
    cache.remove("city");
    std::cout << "After removal, has 'city': " << (cache.has("city") ? "yes" : "no") << std::endl;

    // Get with default
    std::string age = cache.get_or_default("age", std::string("unknown"));
    std::cout << "Age (with default): " << age << std::endl;
}

// Example 2: TTL (Time To Live)
void example_ttl() {
    std::cout << "\n=== Example 2: TTL (Time To Live) ===" << std::endl;

    Cache<std::string, std::string> cache(100);

    // Set with 2 second TTL
    cache.set("session_token", "abc123", 2);
    std::cout << "Session token set with 2 second TTL" << std::endl;
    std::cout << "Has token: " << (cache.has("session_token") ? "yes" : "no") << std::endl;

    // Wait for expiration
    std::cout << "Waiting 2.5 seconds..." << std::endl;
    std::this_thread::sleep_for(std::chrono::milliseconds(2500));

    std::cout << "Has token after wait: " << (cache.has("session_token") ? "yes" : "no") << std::endl;

    // Permanent entry (no TTL)
    cache.set("config", "app_settings");
    std::cout << "Permanent config entry: " << (cache.has("config") ? "yes" : "no") << std::endl;
}

// Example 3: LRU Eviction
void example_lru_eviction() {
    std::cout << "\n=== Example 3: LRU Eviction Policy ===" << std::endl;

    // Cache with max 3 entries, LRU eviction
    Cache<std::string, int> cache(3, EvictionPolicy::LRU);

    cache.set("a", 1);
    cache.set("b", 2);
    cache.set("c", 3);
    std::cout << "Added a, b, c. Size: " << cache.size() << std::endl;

    // Access 'a' to make it most recently used
    int val;
    cache.get("a", val);
    std::cout << "Accessed 'a'" << std::endl;

    // Add new entry - 'b' should be evicted (least recently used)
    cache.set("d", 4);
    std::cout << "Added 'd'. Size: " << cache.size() << std::endl;

    std::cout << "Has 'a': " << (cache.has("a") ? "yes" : "no") << std::endl;
    std::cout << "Has 'b': " << (cache.has("b") ? "yes" : "no") << " (evicted)" << std::endl;
    std::cout << "Has 'c': " << (cache.has("c") ? "yes" : "no") << std::endl;
    std::cout << "Has 'd': " << (cache.has("d") ? "yes" : "no") << std::endl;
}

// Example 4: Statistics tracking
void example_statistics() {
    std::cout << "\n=== Example 4: Statistics Tracking ===" << std::endl;

    Cache<std::string, int> cache(100);

    // Initial stats
    CacheStats stats = cache.stats();
    std::cout << "Initial - Hits: " << stats.hits << ", Misses: " << stats.misses << std::endl;

    // Some misses
    int val;
    cache.get("missing1", val);
    cache.get("missing2", val);

    // Add and hit
    cache.set("key1", 100);
    cache.get("key1", val);
    cache.get("key1", val);

    stats = cache.stats();
    std::cout << "After operations - Hits: " << stats.hits << ", Misses: " << stats.misses << std::endl;
    std::cout << "Hit rate: " << (stats.hit_rate() * 100) << "%" << std::endl;

    // Reset stats
    cache.reset_stats();
    stats = cache.stats();
    std::cout << "After reset - Hits: " << stats.hits << ", Misses: " << stats.misses << std::endl;
}

// Example 5: Get or compute pattern
void example_get_or_compute() {
    std::cout << "\n=== Example 5: Get or Compute Pattern ===" << std::endl;

    Cache<std::string, int> cache(100);

    int compute_count = 0;

    // Function to simulate expensive computation
    // First call - should compute
    std::cout << "First call for 'hello':" << std::endl;
    int val1 = cache.get_or_compute("hello", [&]() {
        compute_count++;
        std::cout << "  Computing value..." << std::endl;
        return 42;
    });
    std::cout << "Value: " << val1 << std::endl;

    // Second call - should use cache
    std::cout << "Second call for 'hello':" << std::endl;
    int val2 = cache.get_or_compute("hello", [&]() {
        compute_count++;
        std::cout << "  Computing value..." << std::endl;
        return 42;
    });
    std::cout << "Value: " << val2 << std::endl;

    std::cout << "Total computations: " << compute_count << std::endl;
}

// Example 6: Different eviction policies
void example_eviction_policies() {
    std::cout << "\n=== Example 6: Eviction Policies ===" << std::endl;

    // FIFO cache
    {
        std::cout << "\nFIFO Policy:" << std::endl;
        Cache<std::string, int> cache(3, EvictionPolicy::FIFO);

        cache.set("a", 1);
        cache.set("b", 2);
        cache.set("c", 3);
        int val;
        cache.get("a", val); // Access doesn't affect FIFO

        cache.set("d", 4); // Evicts 'a' (first in)

        std::cout << "  Has 'a': " << (cache.has("a") ? "yes" : "no") << std::endl;
        std::cout << "  Has 'b': " << (cache.has("b") ? "yes" : "no") << std::endl;
    }

    // LFU cache (Least Frequently Used)
    {
        std::cout << "\nLFU Policy:" << std::endl;
        Cache<std::string, int> cache(3, EvictionPolicy::LFU);

        cache.set("a", 1);
        cache.set("b", 2);
        cache.set("c", 3);

        // Access 'a' multiple times
        int val;
        cache.get("a", val);
        cache.get("a", val);
        cache.get("b", val);

        cache.set("d", 4); // Evicts 'c' (least frequently used)

        std::cout << "  Has 'a': " << (cache.has("a") ? "yes" : "no") << std::endl;
        std::cout << "  Has 'b': " << (cache.has("b") ? "yes" : "no") << std::endl;
        std::cout << "  Has 'c': " << (cache.has("c") ? "yes" : "no") << " (evicted)" << std::endl;
    }
}

// Example 7: Integer keys
void example_integer_keys() {
    std::cout << "\n=== Example 7: Integer Keys ===" << std::endl;

    Cache<int, std::string> cache(100);

    cache.set(1, "one");
    cache.set(2, "two");
    cache.set(3, "three");
    cache.set(100, "hundred");

    std::string val;
    if (cache.get(1, val)) std::cout << "Key 1: " << val << std::endl;
    if (cache.get(2, val)) std::cout << "Key 2: " << val << std::endl;
    if (cache.get(100, val)) std::cout << "Key 100: " << val << std::endl;

    std::vector<int> keys = cache.keys();
    std::cout << "All keys: ";
    for (size_t i = 0; i < keys.size(); i++) {
        std::cout << keys[i] << " ";
    }
    std::cout << std::endl;
}

// Example 8: Thread-safe operations
void example_thread_safety() {
    std::cout << "\n=== Example 8: Thread-Safe Operations ===" << std::endl;

    Cache<int, int> cache(1000);

    std::vector<std::thread> threads;

    // Writer threads
    for (int i = 0; i < 3; i++) {
        threads.push_back(std::thread([&cache, i]() {
            for (int j = 0; j < 50; j++) {
                cache.set(i * 100 + j, j * j);
            }
        }));
    }

    // Reader threads
    for (int i = 0; i < 2; i++) {
        threads.push_back(std::thread([&cache]() {
            int val;
            for (int j = 0; j < 100; j++) {
                cache.get(j, val);
            }
        }));
    }

    for (size_t i = 0; i < threads.size(); i++) {
        threads[i].join();
    }

    std::cout << "Final cache size after concurrent operations: " << cache.size() << std::endl;
    CacheStats stats = cache.stats();
    std::cout << "Hits: " << stats.hits << ", Misses: " << stats.misses << std::endl;
}

int main() {
    std::cout << "========================================" << std::endl;
    std::cout << "Cache Utilities Examples" << std::endl;
    std::cout << "========================================" << std::endl;

    example_basic_usage();
    example_ttl();
    example_lru_eviction();
    example_statistics();
    example_get_or_compute();
    example_eviction_policies();
    example_integer_keys();
    example_thread_safety();

    std::cout << "\n========================================" << std::endl;
    std::cout << "All examples completed successfully!" << std::endl;
    std::cout << "========================================" << std::endl;

    return 0;
}

File: R/examples/cache_utils_example.R
# cache_utils_example.R
# Example usage of R cache_utils module
# Run with: Rscript cache_utils_example.R

source("../cache_utils/mod.R")

cat("R Cache Utilities Examples")
cat("==========================\n\n")

# Example 1: Basic cache operations
cat("1. Basic Cache Operations\n")
cat("---------------------------\n")
cache <- cache_create(max_size = 100)

# Set values
cache$set("user:123", list(name = "Alice", age = 30))
cache$set("user:456", list(name = "Bob", age = 25))

# Get values
user1 <- cache$get("user:123")
cat("User 123:", user1$name, ",", user1$age, "years old\n")

user2 <- cache$get("user:456")
cat("User 456:", user2$name, ",", user2$age, "years old\n")

# Check existence
cat("Has user:123?", cache$has("user:123"), "\n")
cat("Has user:999?", cache$has("user:999"), "\n")

# Get with default
result <- cache$get("user:999", default = list(name = "Unknown"))
cat("User 999 (with default):", result$name, "\n")
cat("\n")

# Example 2: TTL (Time To Live)
cat("2. TTL (Time To Live)\n")
cat("---------------------\n")
cache2 <- cache_create()

# Set with TTL of 2 seconds
cache2$set("session_token", "abc123xyz", ttl = 2)
cat("Session token set with 2 second TTL\n")
cat("Has session_token?", cache2$has("session_token"), "\n")
cat("TTL remaining:", cache2$ttl("session_token"), "seconds\n")

cat("Waiting 3 seconds...\n")
Sys.sleep(3)

cat("Has session_token after expiration?", cache2$has("session_token"), "\n")
cat("\n")

# Example 3: Cache Statistics
cat("3. Cache Statistics\n")
cat("-------------------\n")
cache3 <- cache_create()

# Generate some cache activity
for (i in 1:10) {
  cache3$set(paste0("key", i), paste0("value", i))
}

# Access some keys (hits)
for (i in 1:5) {
  cache3$get(paste0("key", i))
}

# Try to access non-existent keys (misses)
for (i in 11:15) {
  cache3$get(paste0("key", i))
}

stats <- cache3$get_stats()
cat("Cache Statistics:\n")
cat("  Hits:", stats$hits, "\n")
cat("  Misses:", stats$misses, "\n")
cat("  Hit Rate:", round(stats$hit_rate * 100, 2), "%\n")
cat("  Sets:", stats$sets, "\n")
cat("  Current Size:", stats$current_size, "/", stats$max_size, "\n")
cat("\n")

# Example 4: LRU Eviction
cat("4. LRU (Least Recently Used) Eviction\n")
cat("--------------------------------------\n")
cache4 <- cache_create(max_size = 3)

cache4$set("a", "1")
cache4$set("b", "2")
cache4$set("c", "3")
cat("Cache size after 3 inserts:", cache4$size(), "\n")

# Access 'a' to make it recently used
cache4$get("a")

# Add 'd', should evict 'b' (least recently used)
cache4$set("d", "4")
cat("Cache size after adding 'd':", cache4$size(), "\n")
cat("Has 'a'?", cache4$has("a"), "(accessed recently)\n")
cat("Has 'b'?", cache4$has("b"), "(evicted - LRU)\n")
cat("Has 'c'?", cache4$has("c"), "\n")
cat("Has 'd'?", cache4$has("d"), "\n")

stats <- cache4$get_stats()
cat("Evictions:", stats$evictions, "\n")
cat("\n")

# Example 5: Get or Set (Memoization Pattern)
cat("5. Get or Set (Memoization)\n")
cat("----------------------------\n")
cache5 <- cache_create()

# Simulate expensive computation
expensive_function <- function(x) {
  cat("  Computing for", x, "...\n")
  Sys.sleep(0.1)  # Simulate work
  return(x * x)
}

cat("First call (will compute):\n")
result1 <- cache5$get_or_set("square:5", function() expensive_function(5))
cat("Result:", result1, "\n")

cat("Second call (cached, no computation):\n")
result2 <- cache5$get_or_set("square:5", function() expensive_function(5))
cat("Result:", result2, "\n")

cat("Different key (will compute):\n")
result3 <- cache5$get_or_set("square:10", function() expensive_function(10))
cat("Result:", result3, "\n")
cat("\n")

# Example 6: Multiple Operations
cat("6. Multiple Operations\n")
cat("----------------------\n")
cache6 <- cache_create()

# Set multiple values at once
cache6$set_multiple(list(
  product_1 = list(name = "Laptop", price = 999),
  product_2 = list(name = "Mouse", price = 29),
  product_3 = list(name = "Keyboard", price = 79)
))

# Get multiple values at once
results <- cache6$get_multiple(c("product_1", "product_2", "product_99"))
cat("Product 1:", results$product_1$name, "- $", results$product_1$price, "\n")
cat("Product 2:", results$product_2$name, "- $", results$product_2$price, "\n")
cat("Product 99 (not found):", is.null(results$product_99), "\n")

# Delete multiple
deleted <- cache6$delete_multiple(c("product_1", "product_2", "product_99"))
cat("Deleted", deleted, "items\n")
cat("Remaining keys:", paste(cache6$keys(), collapse = ", "), "\n")
cat("\n")

# Example 7: Increment/Decrement
cat("7. Increment/Decrement\n")
cat("----------------------\n")
cache7 <- cache_create()

# Counter operations
cache7$increment("views", 1)
cache7$increment("views", 1)
cache7$increment("views", 1)
cat("Views after 3 increments:", cache7$get("views"), "\n")

cache7$decrement("views", 1)
cat("Views after 1 decrement:", cache7$get("views"), "\n")

# Custom increment amount
cache7$increment("score", 100)
cache7$increment("score", 50)
cat("Score after increments:", cache7$get("score"), "\n")
cat("\n")

# Example 8: Different Data Types
cat("8. Different Data Types\n")
cat("-----------------------\n")
cache8 <- cache_create()

# Store various R data types
cache8$set("string", "Hello, World!")
cache8$set("number", 3.14159)
cache8$set("integer", 42L)
cache8$set("logical", TRUE)
cache8$set("vector", c(1, 2, 3, 4, 5))
cache8$set("matrix", matrix(1:9, nrow = 3))
cache8$set("data_frame", data.frame(
  name = c("Alice", "Bob"),
  age = c(30, 25),
  active = c(TRUE, FALSE)
))
cache8$set("list", list(
  nested = list(
    deeply = list(
      value = "found me!"
    )
  )
))

# Retrieve and display
cat("String:", cache8$get("string"), "\n")
cat("Number:", cache8$get("number"), "\n")
cat("Integer:", cache8$get("integer"), "\n")
cat("Logical:", cache8$get("logical"), "\n")
cat("Vector:", paste(cache8$get("vector"), collapse = ", "), "\n")
cat("Matrix:\n")
print(cache8$get("matrix"))
cat("Data Frame:\n")
print(cache8$get("data_frame"))
nested <- cache8$get("list")
cat("Nested list value:", nested$nested$deeply$value, "\n")
cat("\n")

# Example 9: Cache with Default TTL
cat("9. Cache with Default TTL\n")
cat("--------------------------\n")
cache9 <- cache_create(max_size = 100, default_ttl = 1)  # 1 second default TTL

cache9$set("temp1", "value1")  # Uses default TTL
cache9$set("temp2", "value2", ttl = 5)  # Override with 5 seconds
cache9$set("permanent", "value3", ttl = NULL)  # No expiration

cat("Initial state:\n")
cat("  Has temp1?", cache9$has("temp1"), "\n")
cat("  Has temp2?", cache9$has("temp2"), "\n")
cat("  Has permanent?", cache9$has("permanent"), "\n")

cat("Waiting 1.5 seconds...\n")
Sys.sleep(1.5)

cat("After 1.5 seconds:\n")
cat("  Has temp1 (expired)?", cache9$has("temp1"), "\n")
cat("  Has temp2 (still valid)?", cache9$has("temp2"), "\n")
cat("  Has permanent (never expires)?", cache9$has("permanent"), "\n")
cat("\n")

# Example 10: Practical Use Case - API Response Caching
cat("10. Practical Use Case: API Response Caching\n")
cat("---------------------------------------------\n")

# Simulate an API client with caching
api_cache <- cache_create(max_size = 50, default_ttl = 300)  # 5 minute cache

fetch_user_data <- function(user_id) {
  # Check cache first
  cache_key <- paste0("user:", user_id)
  cached <- api_cache$get(cache_key)
  
  if (!is.null(cached)) {
    cat("  [CACHE HIT] Returning cached data for user", user_id, "\n")
    return(cached)
  }
  
  # Simulate API call
  cat("  [API CALL] Fetching user", user_id, "from API...\n")
  Sys.sleep(0.2)  # Simulate network delay
  
  # Return mock data
  user_data <- list(
    id = user_id,
    name = paste("User", user_id),
    email = paste0("user", user_id, "@example.com"),
    created_at = Sys.time()
  )
  
  # Cache the result
  api_cache$set(cache_key, user_data)
  cat("  [CACHE SET] Stored user", user_id, "in cache\n")
  
  return(user_data)
}

cat("Fetching user 101:\n")
user1 <- fetch_user_data(101)
cat("  Name:", user1$name, "\n\n")

cat("Fetching user 101 again (should be cached):\n")
user1_cached <- fetch_user_data(101)
cat("  Name:", user1_cached$name, "\n\n")

cat("Fetching user 102:\n")
user2 <- fetch_user_data(102)
cat("  Name:", user2$name, "\n\n")

# Show cache stats
stats <- api_cache$get_stats()
cat("API Cache Statistics:\n")
cat("  Hits:", stats$hits, "\n")
cat("  Misses:", stats$misses, "\n")
cat("  Hit Rate:", round(stats$hit_rate * 100, 2), "%\n")
cat("\n")

# Print the cache object (uses custom print method)
cat("Cache object summary:\n")
print(api_cache)

cat("\n==========================\n")
cat("Examples completed!\n")
